File: iotdb-client/isession/src/main/java/org/apache/iotdb/isession/SessionDataSet.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iotdb.isession;

import org.apache.iotdb.rpc.IoTDBConnectionException;
import org.apache.iotdb.rpc.IoTDBRpcDataSet;
import org.apache.iotdb.rpc.RpcUtils;
import org.apache.iotdb.rpc.StatementExecutionException;
import org.apache.iotdb.service.rpc.thrift.IClientRPCService;

import org.apache.thrift.TException;
import org.apache.tsfile.enums.TSDataType;
import org.apache.tsfile.read.common.Field;
import org.apache.tsfile.read.common.RowRecord;
import org.apache.tsfile.write.UnSupportedDataTypeException;

import java.nio.ByteBuffer;
import java.sql.Timestamp;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class SessionDataSet implements ISessionDataSet {

  private final IoTDBRpcDataSet ioTDBRpcDataSet;

  @SuppressWarnings("squid:S107") // ignore Methods should not have too many parameters
  public SessionDataSet(
      String sql,
      List<String> columnNameList,
      List<String> columnTypeList,
      Map<String, Integer> columnNameIndex,
      long queryId,
      long statementId,
      IClientRPCService.Iface client,
      long sessionId,
      List<ByteBuffer> queryResult,
      boolean ignoreTimeStamp,
      boolean moreData,
      ZoneId zoneId,
      int timeFactor,
      boolean tableModel,
      List<Integer> columnIndex2TsBlockColumnIndexList) {
    this.ioTDBRpcDataSet =
        new IoTDBRpcDataSet(
            sql,
            columnNameList,
            columnTypeList,
            columnNameIndex,
            ignoreTimeStamp,
            moreData,
            queryId,
            statementId,
            client,
            sessionId,
            queryResult,
            SessionConfig.DEFAULT_FETCH_SIZE,
            0,
            zoneId,
            RpcUtils.DEFAULT_TIME_FORMAT,
            timeFactor,
            tableModel,
            columnIndex2TsBlockColumnIndexList);
  }

  @SuppressWarnings("squid:S107") // ignore Methods should not have too many parameters
  public SessionDataSet(
      String sql,
      List<String> columnNameList,
      List<String> columnTypeList,
      Map<String, Integer> columnNameIndex,
      long queryId,
      long statementId,
      IClientRPCService.Iface client,
      long sessionId,
      List<ByteBuffer> queryResult,
      boolean ignoreTimeStamp,
      long timeout,
      boolean moreData,
      int fetchSize,
      ZoneId zoneId,
      int timeFactor,
      boolean tableModel,
      List<Integer> columnIndex2TsBlockColumnIndexList) {
    this.ioTDBRpcDataSet =
        new IoTDBRpcDataSet(
            sql,
            columnNameList,
            columnTypeList,
            columnNameIndex,
            ignoreTimeStamp,
            moreData,
            queryId,
            statementId,
            client,
            sessionId,
            queryResult,
            fetchSize,
            timeout,
            zoneId,
            RpcUtils.DEFAULT_TIME_FORMAT,
            timeFactor,
            tableModel,
            columnIndex2TsBlockColumnIndexList);
  }

  public int getFetchSize() {
    return ioTDBRpcDataSet.getFetchSize();
  }

  public void setFetchSize(int fetchSize) {
    ioTDBRpcDataSet.setFetchSize(fetchSize);
  }

  @Override
  public List<String> getColumnNames() {
    return new ArrayList<>(ioTDBRpcDataSet.getColumnNameList());
  }

  @Override
  public List<String> getColumnTypes() {
    return new ArrayList<>(ioTDBRpcDataSet.getColumnTypeList());
  }

  @Override
  public boolean hasNext() throws StatementExecutionException, IoTDBConnectionException {
    if (ioTDBRpcDataSet.hasCachedRecord()) {
      return true;
    } else {
      return ioTDBRpcDataSet.next();
    }
  }

  private RowRecord constructRowRecordFromValueArray() throws StatementExecutionException {
    List<Field> outFields = new ArrayList<>();
    for (int i = ioTDBRpcDataSet.getValueColumnStartIndex();
        i < ioTDBRpcDataSet.getColumnSize();
        i++) {
      Field field;

      String columnName = ioTDBRpcDataSet.getColumnNameList().get(i);

      if (!ioTDBRpcDataSet.isNull(columnName)) {
        TSDataType dataType = ioTDBRpcDataSet.getDataType(columnName);
        field = new Field(dataType);
        switch (dataType) {
          case BOOLEAN:
            boolean booleanValue = ioTDBRpcDataSet.getBoolean(columnName);
            field.setBoolV(booleanValue);
            break;
          case INT32:
          case DATE:
            int intValue = ioTDBRpcDataSet.getInt(columnName);
            field.setIntV(intValue);
            break;
          case INT64:
          case TIMESTAMP:
            long longValue = ioTDBRpcDataSet.getLong(columnName);
            field.setLongV(longValue);
            break;
          case FLOAT:
            float floatValue = ioTDBRpcDataSet.getFloat(columnName);
            field.setFloatV(floatValue);
            break;
          case DOUBLE:
            double doubleValue = ioTDBRpcDataSet.getDouble(columnName);
            field.setDoubleV(doubleValue);
            break;
          case TEXT:
          case BLOB:
          case STRING:
            field.setBinaryV(ioTDBRpcDataSet.getBinary(columnName));
            break;
          default:
            throw new UnSupportedDataTypeException(
                String.format("Data type %s is not supported.", dataType));
        }
      } else {
        field = new Field(null);
      }
      outFields.add(field);
    }
    return new RowRecord(ioTDBRpcDataSet.getCurrentRowTime(), outFields);
  }

  /**
   * Iterate ResultSet using this method isn't very efficient, because it will use RowRecord to
   * represent a row which contains much object creation and converting overhead If you just want to
   * get each value of each column row by row, you can use SessionDataSet.iterator() to get
   * DataIterator, and use DataIterator.getXXX() function to get current row's specified column
   * value.
   *
   * @return One complete row saved in RowRecord
   */
  @Override
  public RowRecord next() throws StatementExecutionException, IoTDBConnectionException {
    if (!ioTDBRpcDataSet.hasCachedRecord() && !hasNext()) {
      return null;
    }
    ioTDBRpcDataSet.setHasCachedRecord(false);

    return constructRowRecordFromValueArray();
  }

  public void closeOperationHandle() throws StatementExecutionException, IoTDBConnectionException {
    try {
      ioTDBRpcDataSet.close();
    } catch (TException e) {
      throw new IoTDBConnectionException(e.getMessage());
    }
  }

  public DataIterator iterator() {
    return new DataIterator();
  }

  @Override
  public void close() throws IoTDBConnectionException, StatementExecutionException {
    closeOperationHandle();
  }

  public class DataIterator {

    public boolean next() throws StatementExecutionException, IoTDBConnectionException {
      return ioTDBRpcDataSet.next();
    }

    public boolean isNull(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.isNull(columnIndex);
    }

    public boolean isNull(String columnName) {
      return ioTDBRpcDataSet.isNull(columnName);
    }

    public boolean getBoolean(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.getBoolean(columnIndex);
    }

    public boolean getBoolean(String columnName) throws StatementExecutionException {
      return ioTDBRpcDataSet.getBoolean(columnName);
    }

    public double getDouble(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.getDouble(columnIndex);
    }

    public double getDouble(String columnName) throws StatementExecutionException {
      return ioTDBRpcDataSet.getDouble(columnName);
    }

    public float getFloat(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.getFloat(columnIndex);
    }

    public float getFloat(String columnName) throws StatementExecutionException {
      return ioTDBRpcDataSet.getFloat(columnName);
    }

    public int getInt(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.getInt(columnIndex);
    }

    public int getInt(String columnName) throws StatementExecutionException {
      return ioTDBRpcDataSet.getInt(columnName);
    }

    public long getLong(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.getLong(columnIndex);
    }

    public long getLong(String columnName) throws StatementExecutionException {
      return ioTDBRpcDataSet.getLong(columnName);
    }

    public Object getObject(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.getObject(columnIndex);
    }

    public Object getObject(String columnName) throws StatementExecutionException {
      return ioTDBRpcDataSet.getObject(columnName);
    }

    public String getString(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.getString(columnIndex);
    }

    public String getString(String columnName) throws StatementExecutionException {
      return ioTDBRpcDataSet.getString(columnName);
    }

    public Timestamp getTimestamp(int columnIndex) throws StatementExecutionException {
      return ioTDBRpcDataSet.getTimestamp(columnIndex);
    }

    public Timestamp getTimestamp(String columnName) throws StatementExecutionException {
      return ioTDBRpcDataSet.getTimestamp(columnName);
    }

    public int findColumn(String columnName) {
      return ioTDBRpcDataSet.findColumn(columnName);
    }

    public List<String> getColumnNameList() {
      return ioTDBRpcDataSet.getColumnNameList();
    }

    public List<String> getColumnTypeList() {
      return ioTDBRpcDataSet.getColumnTypeList();
    }
  }
}

File: iotdb-client/client-cpp/src/main/SessionDataSet.cpp
/**
* Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "SessionDataSet.h"
#include <stdexcept>
#include <boost/date_time/gregorian/gregorian.hpp>

RowRecord::RowRecord(int64_t timestamp) {
    this->timestamp = timestamp;
}

RowRecord::RowRecord(int64_t timestamp, const std::vector<Field>& fields)
    : timestamp(timestamp), fields(fields) {
}

RowRecord::RowRecord(const std::vector<Field>& fields)
    : timestamp(-1), fields(fields) {
}

RowRecord::RowRecord() {
    this->timestamp = -1;
}

void RowRecord::addField(const Field& f) {
    this->fields.push_back(f);
}

std::string RowRecord::toString() {
    std::string ret;
    if (this->timestamp != -1) {
        ret.append(std::to_string(timestamp));
        ret.append("\t");
    }
    for (size_t i = 0; i < fields.size(); i++) {
        if (i != 0) {
            ret.append("\t");
        }
        TSDataType::TSDataType dataType = fields[i].dataType;
        switch (dataType) {
        case TSDataType::BOOLEAN:
            if (!fields[i].boolV.is_initialized()) {
                ret.append("null");
            } else {
                ret.append(fields[i].boolV.value() ? "true" : "false");
            }
            break;
        case TSDataType::INT32:
            if (!fields[i].intV.is_initialized()) {
                ret.append("null");
            } else {
                ret.append(std::to_string(fields[i].intV.value()));
            }
            break;
        case TSDataType::DATE:
            if (!fields[i].dateV.is_initialized()) {
                ret.append("null");
            } else {
                ret.append(boost::gregorian::to_iso_extended_string(fields[i].dateV.value()));
            }
            break;
        case TSDataType::TIMESTAMP:
        case TSDataType::INT64:
            if (!fields[i].longV.is_initialized()) {
                ret.append("null");
            } else {
                ret.append(std::to_string(fields[i].longV.value()));
            }
            break;
        case TSDataType::FLOAT:
            if (!fields[i].floatV.is_initialized()) {
                ret.append("null");
            } else {
                ret.append(std::to_string(fields[i].floatV.value()));
            }
            break;
        case TSDataType::DOUBLE:
            if (!fields[i].doubleV.is_initialized()) {
                ret.append("null");
            } else {
                ret.append(std::to_string(fields[i].doubleV.value()));
            }
            break;
        case TSDataType::BLOB:
        case TSDataType::STRING:
        case TSDataType::TEXT:
            if (!fields[i].stringV.is_initialized()) {
                ret.append("null");
            } else {
                ret.append(fields[i].stringV.value());
            }
            break;
        default:
            break;
        }
    }
    ret.append("\n");
    return ret;
}

bool SessionDataSet::hasNext() {
    if (iotdbRpcDataSet_->hasCachedRecord()) {
        return true;
    }
    return iotdbRpcDataSet_->next();
}

shared_ptr<RowRecord> SessionDataSet::next() {
    if (!iotdbRpcDataSet_->hasCachedRecord() && !hasNext()) {
        return nullptr;
    }
    iotdbRpcDataSet_->setHasCachedRecord(false);

    return constructRowRecordFromValueArray();
}

int SessionDataSet::getFetchSize() {
    return iotdbRpcDataSet_->getFetchSize();
}

void SessionDataSet::setFetchSize(int fetchSize) {
    return iotdbRpcDataSet_->setFetchSize(fetchSize);
}

const std::vector<std::string>& SessionDataSet::getColumnNames() const {
    return iotdbRpcDataSet_->getColumnNameList();
}

const std::vector<std::string>& SessionDataSet::getColumnTypeList() const {
    return iotdbRpcDataSet_->getColumnTypeList();
}

void SessionDataSet::closeOperationHandle(bool forceClose) {
    iotdbRpcDataSet_->close(forceClose);
}

bool SessionDataSet::DataIterator::next() {
    return iotdbRpcDataSet_->next();
}

bool SessionDataSet::DataIterator::isNull(const std::string& columnName) {
    return iotdbRpcDataSet_->isNullByColumnName(columnName);
}

bool SessionDataSet::DataIterator::isNullByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->isNullByIndex(columnIndex);
}

boost::optional<bool> SessionDataSet::DataIterator::getBooleanByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->getBooleanByIndex(columnIndex);
}

boost::optional<bool> SessionDataSet::DataIterator::getBoolean(const std::string& columnName) {
    return iotdbRpcDataSet_->getBoolean(columnName);
}

boost::optional<double> SessionDataSet::DataIterator::getDoubleByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->getDoubleByIndex(columnIndex);
}

boost::optional<double> SessionDataSet::DataIterator::getDouble(const std::string& columnName) {
    return iotdbRpcDataSet_->getDouble(columnName);
}

boost::optional<float> SessionDataSet::DataIterator::getFloatByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->getFloatByIndex(columnIndex);
}

boost::optional<float> SessionDataSet::DataIterator::getFloat(const std::string& columnName) {
    return iotdbRpcDataSet_->getFloat(columnName);
}

boost::optional<int32_t> SessionDataSet::DataIterator::getIntByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->getIntByIndex(columnIndex);
}

boost::optional<int32_t> SessionDataSet::DataIterator::getInt(const std::string& columnName) {
    return iotdbRpcDataSet_->getInt(columnName);
}

boost::optional<int64_t> SessionDataSet::DataIterator::getLongByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->getLongByIndex(columnIndex);
}

boost::optional<int64_t> SessionDataSet::DataIterator::getLong(const std::string& columnName) {
    return iotdbRpcDataSet_->getLong(columnName);
}

boost::optional<std::string> SessionDataSet::DataIterator::getStringByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->getStringByIndex(columnIndex);
}

boost::optional<std::string> SessionDataSet::DataIterator::getString(const std::string& columnName) {
    return iotdbRpcDataSet_->getString(columnName);
}

boost::optional<int64_t> SessionDataSet::DataIterator::getTimestampByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->getTimestampByIndex(columnIndex);
}

boost::optional<int64_t> SessionDataSet::DataIterator::getTimestamp(const std::string& columnName) {
    return iotdbRpcDataSet_->getTimestamp(columnName);
}

boost::optional<boost::gregorian::date> SessionDataSet::DataIterator::getDateByIndex(int32_t columnIndex) {
    return iotdbRpcDataSet_->getDateByIndex(columnIndex);
}

boost::optional<boost::gregorian::date> SessionDataSet::DataIterator::getDate(const std::string& columnName) {
    return iotdbRpcDataSet_->getDate(columnName);
}

int32_t SessionDataSet::DataIterator::findColumn(const std::string& columnName) {
    return iotdbRpcDataSet_->findColumn(columnName);
}

const std::vector<std::string>& SessionDataSet::DataIterator::getColumnNames() const {
    return iotdbRpcDataSet_->getColumnNameList();
}

const std::vector<std::string>& SessionDataSet::DataIterator::getColumnTypeList() const {
    return iotdbRpcDataSet_->getColumnTypeList();
}

shared_ptr<RowRecord> SessionDataSet::constructRowRecordFromValueArray() {
    std::vector<Field> outFields;
    for (int i = iotdbRpcDataSet_->getValueColumnStartIndex(); i < iotdbRpcDataSet_->getColumnSize(); i++) {
        Field field;
        std::string columnName = iotdbRpcDataSet_->getColumnNameList().at(i);
        if (!iotdbRpcDataSet_->isNullByColumnName(columnName)) {
            TSDataType::TSDataType dataType = iotdbRpcDataSet_->getDataType(columnName);
            field.dataType = dataType;
            switch (dataType) {
            case TSDataType::BOOLEAN:
                field.boolV = iotdbRpcDataSet_->getBoolean(columnName);
                break;
            case TSDataType::INT32:
                field.intV = iotdbRpcDataSet_->getInt(columnName);
                break;
            case TSDataType::DATE:
                field.dateV = iotdbRpcDataSet_->getDate(columnName);
                break;
            case TSDataType::INT64:
            case TSDataType::TIMESTAMP:
                field.longV = iotdbRpcDataSet_->getLong(columnName);
                break;
            case TSDataType::FLOAT:
                field.floatV = iotdbRpcDataSet_->getFloat(columnName);
                break;
            case TSDataType::DOUBLE:
                field.doubleV = iotdbRpcDataSet_->getDouble(columnName);
                break;
            case TSDataType::TEXT:
            case TSDataType::BLOB:
            case TSDataType::STRING:
                field.stringV = iotdbRpcDataSet_->getBinary(columnName)->getStringValue();
                break;
            default:
                throw UnSupportedDataTypeException("Data type %s is not supported." + dataType);
            }
        }
        outFields.emplace_back(field);
    }
    return std::make_shared<RowRecord>(iotdbRpcDataSet_->getCurrentRowTime(), outFields);
}
